EasyGYM/UI/ViewModels/LibraryViewModel.swift

```diff
@@ -6,8 +6,111 @@
 //
 
 import Foundation
+import FirebaseAuth
+import FirebaseFirestore
 
 class LibraryViewModel {
     
-
+    var libraryRepo = WorkoutService()
+    var filteredWorkouts: [Workout] = []
+    var workouts: [Workout] = []
+    var isFiltering: Bool = false
+    var favoriteWorkoutIDs: Set<String> = []
+    
+    func filterWorkouts(by criteria: LibraryViewController.filterCriteria){
+        print("Filtering workouts by \(criteria)")
+        switch criteria {
+        case .push:
+            filteredWorkouts = workouts.filter { $0.movementType == .push }
+        case .pull:
+            filteredWorkouts = workouts.filter{ $0.movementType == .pull }
+        case .leg:
+            filteredWorkouts = workouts.filter{ $0.movementType == .leg }
+        }
+        print("Filtered Workouts: \(filteredWorkouts.count)")
+        isFiltering = true
+    }
+    
+    func clearFilter(){
+        isFiltering = false
+        filteredWorkouts.removeAll()
+    }
+    
+    func fetchWorkouts(completion: @escaping () -> Void) {
+        WorkoutService.shared.fetchWorkouts { [weak self] (workouts, error) in
+            if let error = error {
+                print("Error fetching workouts: \(error.localizedDescription)")
+                return
+            }
+            self?.workouts = workouts ?? []
+            print("Fetched Workouts: \(self?.workouts.count ?? 0)")
+            completion()
+        }
+    }
+    
+    func getWorkoutsCount() -> Int {
+        return workouts.count
+    }
+    
+    func getWorkout(at index: Int) -> Workout {
+        return workouts[index]
+    }
+    
+    
+    func fetchFavoriteWorkouts(completion: @escaping () -> Void) {
+        guard let userID = Auth.auth().currentUser?.uid else { return }
+        let db = Firestore.firestore()
+        db.collection("users").document(userID).collection("favorites").getDocuments { (snapshot, error) in
+            if let error = error {
+                print("Error fetching favorites: \(error.localizedDescription)")
+                return
+            }
+            for document in snapshot?.documents ?? [] {
+                self.favoriteWorkoutIDs.insert(document.documentID)
+            }
+        }
+    }
+    
+    func isFavorite(workoutID: String) -> Bool {
+        return favoriteWorkoutIDs.contains(workoutID)
+    }
+    
+    
+    func didTapFavorite(on workout: Workout, completion: @escaping () -> Void) {
+        guard let userID = Auth.auth().currentUser?.uid else { return }
+        let db = Firestore.firestore()
+        let workoutRef = db.collection("users").document(userID).collection("favorites").document(workout.id)
```

Comment:
i would suggest creating a class that holds all of those delete/setData methods. 
You might run into the issue of wanting to addData in two different views, but then you would need to douplicate the code. 
Its therefore good to create class that holds "db" privately and provides these methods publicly. (also prevents accidental access and manipulation of "db" and you don't need to import FirebaseAuth and FirebaseFirestore so often)

you could also use this method to get rid of creating [String:Any] 
```
func writeToFirestore<T: Codable>(collectionName: String, documentID: String? = nil, object: T, completion: @escaping (Result<Void, Error>) -> Void) {
    let db = Firestore.firestore()
    
    do {
        let data = try Firestore.Encoder().encode(object)
        
        if let documentID = documentID {
            // Specify document ID
            db.collection(collectionName).document(documentID).setData(data) { error in
                if let error = error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
        } else {
            // Auto-generate document ID
            db.collection(collectionName).addDocument(data: data) { error in
                if let error = error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
        }
    } catch let error {
        completion(.failure(error))
    }
}

```